### ISSUE TITLE
truncatedScan correction algorithm for laminography(45 degree)

### ISSUE THREAD

**Issue by hws203** [LLNL/LEAP#134]
@kylechampley I tested leapct with 45 degree laminography geometry with truncated off  and on option. 
As you can see below, your truncated correction algorithm may not work correctly at the case of laminography.  
How about your thought(tested by leapct's version 1.24) ?

1. truncated on:

![truncated_on](https://github.com/user-attachments/assets/13c18622-60f6-4092-8c01-49bb1d88180f)

2. truncated off:
![truncated_off](https://github.com/user-attachments/assets/39ccae73-92b5-4a1e-a9ce-9d1adcc3e4ff)

3. other vendor truncated on:
![cera_tuncated_on](https://github.com/user-attachments/assets/fdf75865-86b5-450c-87d5-f62ac203ae39)

**Comment by kylechampley** [comment]
It looks to me like there is something wrong with the geometry specification in LEAP.  Are you sure it is correct?  You can also expand the reconstruction field of view in LEAP to match the CERA reconstruction.

**Comment by hws203** [comment]
@kylechampley Yes, the geometry is correct to my machine design spec. And it is focused on the FoV of  slice, then there is no such strains. The strains only exists on the outside boundary of  FOV with truncated on option.

**Comment by hws203** [comment]
@kylechampley I maded this issue at your standard d32_laminography.py file with below modifcations.

1. sod = 1100/2 # source-to-object distance (mm)
2. leapct.set_diameterFOV(1.0e7)
3. leapct.set_truncatedScan(True)
just three changes will show this issue at your lastest v1.24's laminography demo script file.

![strains_oct](https://github.com/user-attachments/assets/b089e702-aa5d-4f68-99ad-c373915ab842)

**Comment by kylechampley** [comment]
I don't know.  The truncated scan option seems to be doing what is it supposed to when I run the code.  Note that all this parameter does is this:  If truncatedScan is False, then zero-padding is done with the ramp filter.  If truncatedScan is True, then the edge 100 values are padded with zero-th order extrapolation and the rest are zeros.

**Comment by hws203** [comment]
[d32_laminography_hws203.zip](https://github.com/user-attachments/files/17669698/d32_laminography_hws203.zip)


Above is my test script zipped file. Please check this one.

**Comment by hws203** [comment]
@kylechampley  I changed the z-axis laminography position calculation method from your original demo script at 50 degree.

1. original method : 
  1-1. sourcePositions[:,2] = np.tan(laminographyAngle*np.pi/180.0)*sod
  1-2. moduleCenters[:,2] = -np.tan(laminographyAngle*np.pi/180.0)*(sdd-sod)
2. original method : 
  2-1. sourcePositions[:,2] = np.sin(laminographyAngle*np.pi/180.0)*sod
  2-2. moduleCenters[:,2] = -np.sin(laminographyAngle*np.pi/180.0)*(sdd-sod)

3. original lami-slice:
 
![lami_50_degree](https://github.com/user-attachments/assets/1a262a7d-a61b-4dc8-8cf9-f09138ef554f)

4. changed lami-slice:
![lami_sin_50_degree](https://github.com/user-attachments/assets/6750e647-5f7f-40e9-baab-1441ee9e0b32)


5. If you use tangent function, then the original FID will be changed. How about your thought?

**Comment by kylechampley** [comment]
I looked at the script you sent, but I didn't see anything obviously wrong.

The truncated scan setting seems to be doing what it was intended to do from my perspective.  Yes, there are better ways to handle truncation aritfacts; these are usually two-pass methods which could easily be done with the LEAP toolbox, but I don't have the time to implement these at this time.

Also note that laminography is an incomplete problem.  By scanning at an angle there are certain frequencies in the reconstruction that cannot be recovered and you will always get artifact- which are similar in appearance and reason as the so-called cone-beam artifact.

**Comment by hws203** [comment]
@kylechampley Thanks for your comments.  current issue has two kind of issues, one is truncated artifact and the other is angle calculation method of laminography  of leapct.  I will close this issue which is related to the first one. Please consider the second one more if you have time.

**Comment by hws203** [comment]
1. 60 degree laminography with Leapct's FDK:
![strains_leapct](https://github.com/user-attachments/assets/5b749f34-7bff-4fbe-a75c-f95fcc624c1a)

2. 60 degree laminography with Xrecon4.0's FDK:
![strains_no](https://github.com/user-attachments/assets/e756f084-b6b1-458d-9dcf-237cf2eedd30)

**Comment by hws203** [comment]
@kylechampley I guess that this issue is related to leapct's geometry bug in the case of laminography.  As you may know, in the case of standard cone beam, the zero degree and 180  degree angle have same FID change slope from detector's center to detector's out boundary but in the case of lanminography, the zero degree and 180  degree angle have not same FID change slope from detector's center to detector's out boundary.  It should be considered by the laminograhy angle factor too.

**Comment by hws203** [comment]
https://github.com/tomography/tomocupy/blob/main/src/include/kernels_linerec.cuh
Tomocupy could be useful reference for laminograph, there are two kind backprojection kernels, one is normal and  the other is for laminography.

**Comment by kylechampley** [comment]
I have not done laminography in a long time and don't plan to look into it in the near future, so I don't know why you are having issues.

My best guess is that you have something incorrectly specified in the geometry.  You are also violating the LEAP assumptions by bypassing the 5 degree detector rotation limitation in the FDK reconstruction.

Have you tried the VD backprojector here?

**Comment by hws203** [comment]
@kylechampley I will check vd back-project tomorrow. And I am adding laminography cone beam kernel into my c++ leapct class. I will post my result soon.

**Comment by hws203** [comment]
I have checked the VD_backprojector, and I can see the same strains on the outsider boundary.  So I needs to keep on my work for laminography in native c++ leapct class.  I have done three functions for laminography.
1.cv::Mat rotation_vec_to_z(Mat rot_dir) => Z-axis based rotation matrix generation from the input of rotation_director_vector(3x1).
1. cv::Mat rotation_vec_to_z(Mat rot_dir) => Z-axis based rotation matrix generation from the input of rotation_director_vector(3x1).
2. cv::Mat rotation_vec_to_y(Mat src_det) => Y-axis based rotation matrix generation from the input of source_to_detector_vector(3x1).
3. std::vector<cv::Mat> align_reference_frame(int bin, int crop, float panel_pitch, int det_width, int det_height,  Mat rotPos, Mat rotDir, Mat source_position, Mat detector_position, Mat detector_direction_x, Mat detector_direction_y) => This function makes the aligned laminography geometry vectors for leapct class's srcVecs, detVecs, rowVecs, colVecs.

I finished above three functions and verified it by comparing the CIL package's output with same input params.

And after updating those geometry vectors into leapct's vectors, I can see backproject_Joseph_modular() works without any error, but the result slice are almost all zeros.

So I needs to finish the remains work to link my vectors to the  ASTRA class for its cone_bp.cu function. 
I already checked the ASTRA's cone_bp.cu makes no strains on slice with CIL package in python.
